File: lib/puppeteerInstagram.js
import { faker } from '@faker-js/faker'
import puppeteer from 'puppeteer-extra'

import { signup } from './signup.js'
import { signin } from './signin.js'
import { signout } from './signout.js'
import { verifyEmail } from './verify-email.js'
import { countryList } from './countryIdList.js'
import { followBack } from './followBack.js'

import puppeteerExtraPluginStealth from 'puppeteer-extra-plugin-stealth'

puppeteer.use(puppeteerExtraPluginStealth())

class PuppeteerInstagram {
  constructor (ws, proxy) {
    this._user = null
    this._ws = ws
    this._proxy = proxy
  }

  get isAuthenticated () {
    return !!this._user
  }

  get user () {
    return this._user
  }

  async browser () {
    if (!this._browser) {
      this._browser = await puppeteer.connect({
        browserWSEndpoint: this._ws,
        defaultViewport: null
      })
      await this._browser.authenticate(this._proxy.username, this._proxy.password)
    }
    return this._browser
  }

  async followBack () {
    // if (!this.isAuthenticated) {
    //   throw new Error('"signout" requires authentication')
    // }
    const browser = await this.browser()
    await followBack(browser)
  }

  async signup (profileId) {
    let CountryId = '15'
    const user = {}
    user.username = faker.internet.userName()
    user.password = faker.internet.password()
    user.firstName = faker.person.firstName()
    user.lastName = faker.person.lastName()
    user.username = user.username
      .trim()
      .toLowerCase()
      .replace(/[^\d\w-]/g, '-')
      .replace(/_/g, '-')
      .replace(/^-/g, '')
      .replace(/-$/g, '')
      .replace(/--/g, '-')
      .replace(/-/, '')
    const SmSPoolAPI = 'https://api.smspool.net/purchase/sms'
    const key = 'hFXGJFunoIckg01PLNJlEqHG5IcG8niv'
    countryList.forEach(element => {
      if (element.name === 'United Kingdom') {
        CountryId = element.ID
      }
    })
    // code 15 is for India  https://api.smspool.net/country/retrieve_all`
    const ServiceId = '457'
    const response = await fetch(`${SmSPoolAPI}?key=${key}&country=${CountryId}&service=${ServiceId}`)
    const jsonData = await response.json()
    const phoneNumber = jsonData.phonenumber
    const orderId = jsonData.order_id
    const country = jsonData.country
    const success = jsonData.success
    const countryCode = jsonData.cc
    const message = jsonData.message
    user.number = '+' + String(countryCode) + String(phoneNumber)
    user.orderId = orderId
    user.key = key
    console.log('[UserInformation]', { phoneNumber: user.number, orderId, country, success, countryCode, message })
    if (message.startsWith('This country is currently not available for this service')) {
      console.error('[Error Message]', { jsonData })
    }
    const browser = await this.browser()
    await signup(browser, user, profileId)
  }

  async signin (user, opts = {}) {
    if (this.isAuthenticated) {
      throw new Error('"signin" requires no authentication')
    }

    const browser = await this.browser()
    await signin(browser, user, opts)

    this._user = user
  }

  async signout () {
    if (!this.isAuthenticated) {
      throw new Error('"signout" requires authentication')
    }
    const browser = await this.browser()

    await signout(browser, this._user)
    this._user = null
  }

  async verifyEmail (opts) {
    const browser = await this.browser()
    await verifyEmail(browser, {
      email: opts.email || this.user.email,
      password: opts.emailPassword
    }, opts)
  }

  async close () {
    const browser = await this.browser()
    await browser.close()

    this._browser = null
    this._user = null
  }
}

export default PuppeteerInstagram

File: lib/puppeteerInstagram.py
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
import re
import requests
from faker import Faker
from signup import signup

faker = Faker()
class PuppeteerInstagram:
    def __init__(self):
        self._user = None
        # self._ws = ws
        self._browser = None
    
    @property
    def isAuthenticated(self):
        return bool(self._user)
    
    @property
    def user(self):
        return self._user
    
    async def browser(self):
        if not self._browser:
            options = Options()
            options.add_argument("--headless")  # Optional: Run in headless mode
            options.add_argument("--no-sandbox")
            options.add_argument("--disable-gpu")
            
            # service = Service('/path/to/chromedriver')  # Replace with the actual path to chromedriver executable
            
            # self._browser = webdriver.Chrome(service=service, options=options)
        self._browser = webdriver.Chrome()
        return self._browser
    
    def signup(self):
        CountryId = '15'
        user = {}
        user['username'] = faker.user_name()
        user['password'] = faker.password()
        user['firstName'] = faker.first_name()
        user['lastName'] = faker.last_name()
        user['username'] = re.sub(r'[^\d\w-]', '-', user['username'])
        user['username'] = re.sub(r'_', '-', user['username'])
        user['username'] = re.sub(r'^-', '', user['username'])
        user['username'] = re.sub(r'-$', '', user['username'])
        user['username'] = re.sub(r'--', '-', user['username'])
        user['username'] = re.sub(r'-', '', user['username'])
        SmSPoolAPI = 'https://api.smspool.net/purchase/sms'
        key = 'hFXGJFunoIckg01PLNJlEqHG5IcG8niv'
        # for element in countryList:
        #     if element['name'] == 'United Kingdom':
        #         CountryId = element['ID']
        ServiceId = '457'
        response = requests.get(f"{SmSPoolAPI}?key={key}&country={CountryId}&service={ServiceId}")
        jsonData = response.json()
        phoneNumber = jsonData['phonenumber']
        orderId = jsonData['order_id']
        country = jsonData['country']
        success = jsonData['success']
        countryCode = jsonData['cc']
        message = jsonData['message']
        user['number'] = '+' + str(countryCode) + str(phoneNumber)
        user['orderId'] = orderId
        user['key'] = key
        print('[UserInformation]', {'phoneNumber': user['number'], 'orderId': orderId, 'country': country, 'success': success, 'countryCode': countryCode, 'message': message})
        if message.startswith('This country is currently not available for this service'):
            print('[Error Message]', {'jsonData': jsonData})
        browser= webdriver.Chrome('/Users/anmolagarwal/Downloads/driver_path/chromedriver')
        signup(browser, user)

    def close(self):
        browser = self.get_browser()
        browser.close()
        self._browser = None
        self._user = None

# instagram = SeleniumInstagram(ws, proxy)
profileId="ahdjbd"
instagram = PuppeteerInstagram()
instagram.signup()